Lib/asyncio/runners.py

```diff
@@ -97,7 +103,21 @@ def run(self, coro, *, context=None):
 
         if context is None:
             context = self._context
-        task = self._loop.create_task(coro, context=context)
+
+        if futures.isfuture(coro):
```

Comment:
You replace most of this with call to ensure_future.

Reply:
That's what I did originally, but it caused problems with some of the unit tests. In particular, ensure_future() had a call to get_event_loop() in it which caused a unit test to fail. Also, in my original change, I needed to modify ensure_future() to take a context argument and that changed a public API, which caused other associated unit test issues. More importantly, ensure_future() also had code which called close() on the coro or future, which isn't wanted here.

After trying out various alternatives, I settled on just taking the few lines from ensure_future() that were needed and modifying them to include only what this new code required. It was a smaller change, and in my opinion lower risk and more readable.

Reply:
Maybe extract it to a new function? I really don't want to repeat the logic of awaitable trick here.

Reply:
I'll take another look at this tonight.

Reply:
Ok - I figured out a way to leverage the code in ensure_future() for this - see commit 4ed7357.

One thing I'm not thrilled about here is that this adds a new optional keyword argument to `ensure_future()`, which is a public API. I could move the entire logic of `ensure_future()` into a new function instead to avoid changing its signature, but that's kind of ugly in a different way.

One other thing to note is that this now raises a `TypeError` instead of a `ValueError` when a caller passes in a `coro` which is not one of the expected types, since that's what `ensure_future()` raised.